src/debuginfo.rs

```diff
@@ -1,41 +1,170 @@
+    debug_context: &mut FunctionDebugContext<'tcx, (), Location<'gcc>>,
+    instantiated: &mut BitSet<SourceScope>,
+    scope: SourceScope,
+) {
+    if instantiated.contains(scope) {
+        return;
+    }
+
+    let scope_data = &mir.source_scopes[scope];
+    let parent_scope = if let Some(parent) = scope_data.parent_scope {
+        make_mir_scope(cx, instance, mir, variables, debug_context, instantiated, parent);
+        debug_context.scopes[parent]
+    } else {
+        // The root is the function itself.
+        let file = cx.sess().source_map().lookup_source_file(mir.span.lo());
+        debug_context.scopes[scope] = DebugScope {
+            file_start_pos: file.start_pos,
+            file_end_pos: file.end_position(),
+            ..debug_context.scopes[scope]
+        };
+        instantiated.insert(scope);
+        return;
+    };
+
+    if let Some(vars) = variables
+    {
+        if !vars.contains(scope)
+            && scope_data.inlined.is_none() {
+                // Do not create a DIScope if there are no variables defined in this
+                // MIR `SourceScope`, and it's not `inlined`, to avoid debuginfo bloat.
+                debug_context.scopes[scope] = parent_scope;
+                instantiated.insert(scope);
+                return;
+            }
     }
 
-    fn set_dbg_loc(&mut self, _dbg_loc: Self::DILocation) {
-        unimplemented!();
+    let loc = cx.lookup_debug_loc(scope_data.span.lo());
+    let dbg_scope = ();
+
+    let inlined_at = scope_data.inlined.map(|(_, callsite_span)| {
+        // FIXME(eddyb) this doesn't account for the macro-related
+        // `Span` fixups that `rustc_codegen_ssa::mir::debuginfo` does.
+        let callsite_scope = parent_scope.adjust_dbg_scope_for_span(cx, callsite_span);
+        cx.dbg_loc(callsite_scope, parent_scope.inlined_at, callsite_span)
+    });
+    let p_inlined_at = parent_scope.inlined_at;
+    // TODO(tempdragon): dbg_scope: Add support for scope extension here.
+    inlined_at.or(p_inlined_at);
+    
+    debug_context.scopes[scope] = DebugScope {
+        dbg_scope,
+        inlined_at,
+        file_start_pos: loc.0.start_pos,
+        file_end_pos: loc.0.end_position(),
+    };
+    instantiated.insert(scope);
+}
+
+/// Copied from LLVM backend
+impl<'gcc, 'tcx> CodegenCx<'gcc, 'tcx> {
+    pub fn lookup_debug_loc(&self, pos: BytePos) -> (Lrc<SourceFile>, u32, u32) {
+        match self.sess().source_map().lookup_line(pos) {
+            Ok(SourceFileAndLine { sf: file, line }) => {
+                let line_pos = file.lines()[line];
+
+                // Use 1-based indexing.
+                let line = (line + 1) as u32;
+                let col = (file.relative_position(pos) - line_pos).to_u32() + 1;
+                (file,
+                 line,
+                 if ! self.sess().target.is_like_msvc {
+                     col } else {
+                     UNKNOWN_COLUMN_NUMBER
+                 }
```

Comment:
Well, the code is actually not completely portable. Since the scope part is not currently supported(and I haven't found the corresponding structure & interface for a scope in GCC),  I have to currently separate the code from the original LLVM code. I never made it to a correct scope implementation. This will be fixed when the scope support is ready, or maybe never, if it is proved unnecessary.

Reply:
In ef158f295ecd37413a11b9f01d203f7b2440cbe6, I added the original code but commented it as it is currently not in use for an empty implementation of `DIScope`.

Reply:
I'm not sure I understand the difference compared with the code from cg_llvm. It just seems to me that the condition `is_like_msvc` was moved from after the `match` to inside it.

And that the return type was changed from a struct to a tuple containing the same information. Could you please explain to me how this is different?

Reply:
Short answer: It's not that different. I will switch back to the original implementation.

Reply:
I originally wrote this stuff by hand and end up noticing the function being reusable. So I switched to their implementation. I choose a tuple in place of a struct because it seems to me that it is easier to be done not defining a data structure for something that is used only a few times internally.

Reply:
Yeah, I plan to attempt to deduplicate some code between cg_gcc and cg_llvm in the future. Not sure if this specific function can be deduplicated, but I guess a couple a other functions are.

Reply:
Maybe this should be a good idea to do it upstream, but it also depends on the attitude of developers of other backends such as those of cranelift.

Reply:
Yes, it's the plan. Cranelift doesn't use the SSA traits as far as I know.